### PR TITLE
Change `Coqlib.lib_ref` to fail with `NameNotRegistered` instead of `UserError`

### PR DESCRIPTION
**Pull request by LasseBlaauwbroek**
The original motivation for this was that if `core.identity.type` is registered, but `core.identity.sym` not yet, then the `symmetry` tactic will throw a `UserError`. Fixing this would entail catching the exception (which is also done in e.g. `inv.ml`), but it seems that `lib_ref` should not be throwing a `UserError` in the first place.

The `symmetry` bug mentioned above can be triggered by, for example, using `symmetry` on line 35 of `Logic_Type.v`.

**Kind:** bug fix.

- [ ] Added / updated test-suite

### CODE REVIEW

File: tactics/inv.ml [hunk 1]
```diff
@@ -118,9 +118,9 @@ let make_inv_predicate env evd indf realargs id status concl =
    In any case, we carry along the rest of pairs *)
   let eqdata =
     try Coqlib.build_coq_eq_data ()
-    with UserError _ ->
+    with Coqlib.NameNotRegistered _ ->
```
Comment (herbelin) [inline tactics/inv.ml:121]:
```suggestion
    with Coqlib.NameNotRegistered s ->
```
(see below)

File: tactics/inv.ml [hunk 2]
```diff
@@ -118,9 +118,9 @@ let make_inv_predicate env evd indf realargs id status concl =
...
     try Coqlib.build_coq_identity_data ()
-    with UserError _ -> user_err (str "No registered equality.") in
+    with Coqlib.NameNotRegistered _ -> user_err (str "No registered equality.") in
```
Comment (herbelin) [inline tactics/inv.ml:123]: The message could be more informative here:
```suggestion
    with Coqlib.NameNotRegistered _ -> user_err (str "Missing registration of " ++ str s ++ str ".") in
```
(with the `s` coming from the previous failure)

**Review by herbelin — COMMENTED** [review]
It seems also better to me to have a proper exception. When `build_coq_eq_data` fails, telling which component is missing would improve the error message (see suggestion).

I don't know how much existing code would assume a `UserError` already. The change would have to be documented in `dev/doc/changes.md`.

**Comment by LasseBlaauwbroek** [comment]
I´ve added a changelog entry.

**Comment by ejgallego** [comment]
Hi @LasseBlaauwbroek , indeed this is a problem the API has, let's say that the migration to remove hard-coded constants was quite tricky and still work in progress.

The discussion on the API is super-long, so I'll try to summarize what I remember:
- the choice to have `lib_ref` emit a "fatal" error was done on purpose, due to some subtle semantics of exception handling 
- then, users of the API are supposed to call `has_ref` first, and act on `false` before calling `lib_ref`, instead of catching an exception

The other alternative was to make `lib_ref` return an option type,  but this had a huge impact in terms of code.

Thus in my opinion, it is a safer fix to keep the current `user_err` and add instead the required `has_ref` check, as adding a new exception requires a lot of understanding of who is catching it, adding a printer, etc...

**Comment by LasseBlaauwbroek** [comment]
> then, users of the API are supposed to call `has_ref` first

I considered doing this, but for code that calls functions like `Coqlib.build_coq_identity_data` this would entail calling `has_ref` seven times. That´s not very streamlined... It also kind of breaks the encapsulation that `build_coq_identity_data` brings since the calling function has to know exactly what strings are being passed to `lib_ref`.

The easy solution is to keep the `UserError`, and have `Hipattern.match_with_equation` catch it, just like is already happening in `inv.ml` currently. But as I see it catching `UserError` is a recipe for disaster.

**Comment by LasseBlaauwbroek** [comment]
(Note also that `CErrors` does not classify `UserError` as fatal, so it is not clear to me what the advantage of `UserError` over another error is)

**Comment by ejgallego** [comment]
> (Note also that `CErrors` does not classify `UserError` as fatal, so it is not clear to me what the advantage of `UserError` over another error is)

It is not enforced yet, but indeed there is a plan to make it uncatchable, see #11006 , #11002 , and #12602 .

I dunno what'd be best. If you wanna go this route you need to add an exception printer , but given the diff, IMHO it is much simpler just to add the corresponding check in `hipattern` IIUC.

**Comment by herbelin** [comment]
I'm ok to follow @ejgallego's advices.

Actually, it is not clear that building a full `build_coq_eq_data()` is the optimal API to be called by `symmetry` and `transitivity` since, after all, they only require the fields `sym` and `trans` respectively. Some simplifications could be done.

----

Then, this is a longer-term question, but there are several concurrent paths for the `symmetry`/`transitivity` tactic:
1. registration as `core.eq.sym`/`core.eq.trans` to be used when the default equality is found (resp. `core.identity`/`core.JMeq`)
2. registration as a class which `Rewrite.setoid_symmetry` and `Rewrite.setoid_transitivity` can use
3. recognizing of the structural form of the relation with automatic building of an anonymous symmetry/transitivity proof (`Tactics.prove_transitivity` and `Tactics.prove_symmetry`)

Additionally, there is code for canonically generating canonically-named symmetry proofs in `Eqschemes` for equality-like types (currently used in `rewrite` and at inductive time declaration when `Set Rewriting Schemes` is on).

I would say that there is no reason per se to go through `core.eq.sym`/`core.eq.trans` for `symmetry`/`transitivity` (except for compatibility). Afaics, the two mechanisms which are interesting from a user point of view are:
- manual declaration, in which case the class mechanism should be enough (that would mean moving the symmetry/transitivity classes earlier in the prelude, but this is so central, that it may be worth; don't know what @mattam82 would say);
- automatic building of a symmetry/transitivity proof, whether it is named (and registered in a table, e.g. again as a class) or anonymous, and if named, automatically generated at inductive declaration time or generated on a need basis.

**Comment by mattam82** [comment]
I would be fine with symmetry/transitivity being based on the Symmetric/Transitivity typeclasses early in the prelude (RelationClasses seems central in almost any development). I don't think automatic generation of symmetry/transitivity proofs is worth it. Let the user give the proof, and just have default instances for the relations in the code. A user will be fine if we warn: `Could not find an instance of Symmetric R, please declare an instance`.

**Comment by herbelin** [comment]
Thanks for the quick reply! Will there be a ceps about it or something like that?

About "automatic generation of symmetry/transitivity proofs", some code is already there, so I would say that it can also be seen as some user comfort, like the existing automatic generation of various schemes.

**Comment by ejgallego** [comment]
By the way @LasseBlaauwbroek it would be great if we could improve the current API, just making this point clear in case you interpret my suggestions as "let's keep this API". On the contrary, it'd be great to refine it, and many suggestions have been made to that regard.

My comment about being "more conservative" is just from the point of view of doing a minimally invasive quick fix, but sure a deeper revamp of the API would be awesome; as I mentioned the current status is not satisfactory as we had to do a compromise in terms of doing a first step towards removing the hardcoded Gallina names.

**Comment by LasseBlaauwbroek** [comment]
> My comment about being "more conservative" is just from the point of view of doing a minimally invasive quick fix

Sure, I get it. it seems I have inadvertently stumbled into some code that had deeper problems than I thought... For now I´ve changed this PR to just catch the `UserError` like already happens in `Inv` and other places. Sounds to me that a deeper fix is not appropriate right now.

**Review by ejgallego — APPROVED** [review]
That code needs to be rewritten but the current workaround is fine IMO as a hotfix.

**Comment by ejgallego** [comment]
@coqbot: merge now

**Pull request merged**